endpoint/endpoint.go

```diff
@@ -343,3 +343,22 @@ type DNSEndpointList struct {
 	metav1.ListMeta `json:"metadata,omitempty"`
 	Items           []DNSEndpoint `json:"items"`
 }
+
+// Apply filter based on EndpointKey (using DNSName, RecordType & SetIdentifier)
+func RemoveDuplicates(filtered []*Endpoint) []*Endpoint {
```

Comment:
Nitpick: I think `filtered` is a misleading name for this argument. `FilterEndpointsByOwnerID()` above uses the name `eps`, which I also dislike, but maybe it would be good to match.

My personal preference would be `endpoints`.

```suggestion
func RemoveDuplicates(endpoints []*Endpoint) []*Endpoint {
```

Also, I think this is a great opportunity to refactor `dedupSource.Endpoints()`, something like the following.

```go
// Endpoints collects endpoints from its wrapped source and returns them without duplicates.
func (ms *dedupSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error) {
	endpoints, err := ms.source.Endpoints(ctx)
	if err != nil {
		return nil, err
	}

	result := endpoint.RemoveDuplicates(endpoints)

	return result, nil
}
```